Newmac calibration scumonlymac

filmak opened the pull request:
Features of this PR:

- MAC for multi-hop scum networking
- functions for toggling all 16 GPIOs
- skeleton for a scum-only network (currently only two scum)

filmak committed:
add first implementation of multi-hop mac for scum

filmak committed:
add all 16 gpios to be set, cleared, or toggled.

filmak committed:
enable interrupt by ID and removed ICER from disable interrupt by ID

filmak committed:
minor radio.c overhaul, see details

turn the divider OFF by default
add gpio 12-15 toggles for debugging radio
change timer period from 4000 to 2000 (superficial)
added a "receive indefinitely" function to replace the old "receive"
function
made receive callback: not turn RX off and accept all length packets
add rfOn that resets RF controller

filmak committed:
network of ONLY SCUM - pairwise single channel search

filmak committed:
slight modifications to channel_cal code, I believe it still does not work

filmak committed:
added packet index to print statement of freq_sweep_tx

filmak committed:
allow freq_sweep_rx_simple and freq_sweep_rx to receive arbitrary pkt len

daniellovell commented on scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.c:
```diff
@@ -52,4 +53,6 @@ int main(void) {
 //=========================== private =========================================
 void fill_tx_packet(uint8_t *packet, uint8_t packet_len, repeat_rx_tx_state_t state) {
     sprintf(packet, "%d %d %d", state.cfg_coarse, state.cfg_mid, state.cfg_fine);
+		packet[0] = app_vars.index;
```
Fil says this is not necessary

filmak replied:
Yes, reverting this change is fine. I was using the index to do PDR testing. What I should really do is make a new project explicitly for PDR testing.

daniellovell commented on scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.uvprojx:
```diff
@@ -10,12 +10,13 @@
           <Device>ARMCM0</Device>
           <Vendor>ARM</Vendor>
-          <PackID>ARM.CMSIS.4.5.0</PackID>
+          <PackID>ARM.CMSIS.5.7.0</PackID>
```
Will review if this change causes is issues...

daniellovell replied:
This is fine as far as I can tell in my testing. My Keil automatically changes the pack ID to ARM.CMSIS.5.3.0, so I had no issues building.

daniellovell commented on scm_v3c/radio.c:
```diff
@@ -25,7 +25,7 @@ signed short cdr_tau_history[11] = {0};
 //=========================== definition ======================================
 
-#define DIV_ON
+//#define DIV_ON
```
This should become a configuration parameter of the radio init, #def probably not the best way

daniellovell commented on scm_v3c/radio.c:
```diff
@@ -126,6 +129,7 @@ void        build_TX_channel_table(
 
 // pkt_len should include CRC bytes (add 2 bytes to desired pkt size)
 void send_packet(uint8_t *packet, uint8_t pkt_len) {
+		gpio_12_set();
```
Remove GPIO toggles inside radio -Fil

daniellovell replied:
Done in 8cc73b30b98c561e0eee2eb3df5beb06b6cf01e1

daniellovell commented on scm_v3c/radio.c:
```diff
@@ -179,20 +196,26 @@ void cb_endFrame_rx_radio(uint32_t timestamp){
         radio_vars.LQI_chip_errors    = radio_getLQIchipErrors();
-        
         radio_vars.radio_rx_cb(radio_vars.rxPacket, packet_len);
+				gpio_14_toggle();
```
Remove

daniellovell commented on scm_v3c/radio.c:
```diff
@@ -151,8 +156,19 @@ void receive_packet(uint8_t pkt_len) {
+		//printf("rftimer released\r\n");
+}
+
+void receive_packet_indefinitely(uint8_t pkt_len) {
```
Document that its a blocking wait for a packet

daniellovell commented on scm_v3c/rftimer.c:
```diff
@@ -113,7 +113,7 @@ void rftimer_disable_interrupts(void){
     // disable compare interrupt
     *RF_TIMER_REG_CONTROL_ADDRESES[id] = 0x0000;
-    ICER = 0x80;
+    //ICER = 0x80;
```
Is this ok? Needs testing

daniellovell replied:
Addressed by a0b02cebde38e6f564c9355ee45211c3c370a974

filmak replied:
These changes do not seem to influence the operation of my protocols. I'm not sure if it's necessary, but I individually enable_by_id each RFTimer compare interrupt.

daniellovell commented on scm_v3c/scm3c_hw_interface.c:
```diff
@@ -14,8 +14,9 @@
 #define DAC_2M_SETTING_LEN          5
 
 // initialized value for frequency configuration
-#define INIT_HF_CLOCK_FINE          17
```
Remove changes of board-specific values

daniellovell replied:
Fixed in e6eeb40567b76b6e8c1851f27788de37049cf0a9

daniellovell commented on scm_v3c/radio.c:
```diff
@@ -179,20 +196,26 @@ void cb_endFrame_rx_radio(uint32_t timestamp){
-    radio_rfOff();
-    
-    if(packet_len == radio_vars.rxPacket_len && radio_getCrcOk()) {
+    //if(packet_len == radio_vars.rxPacket_len && radio_getCrcOk()) {
```
Proposal to have receive_packet_length(uint8_t packet_len) and receive_packet(void) where the latter does not know a priori the length to expect and does not check it.

daniellovell replied:
Implemented in dd1e918347eb80e30ab1b1c1f82230a4d2b77e00

filmak replied:
In my protocols I effectively rewrite "receive packet" and the "endframe_cb" so any changes are fine. I believe that many of the radio example applications do use these. if I'm understanding the changes correctly, they should work.

daniellovell commented on scm_v3c/applications/freq_sweep_rx/freq_sweep_rx.c:
```diff
@@ -180,7 +180,7 @@ void    cb_endFrame_rx(uint32_t timestamp){
     radio_rfOff();
     
     if(
-        app_vars.packet_len == LEN_RX_PKT && (radio_getCrcOk())
```
Restore packet_len check

daniellovell committed:
Reverted changes to default frequency configuration

daniellovell committed:
Remove extra gpio toggles inside radio.c

daniellovell committed:
Redefined rftimer interrupt enable/disable functions

rftimer_enable/disable_interrupts will toggle on/off ALL interrupts by setting the ISER and ICER
rftimer_enable/disable_interrupts_by_id will not affect the RFTIMER interrupt (0x80) but will only modify the individual compare registers.

daniellovell committed:
Introduced receive_packet with optional timeout and packet length

daniellovell committed:
Restore state of modified applications
